bob-tools/bin/config_cluster_generator/center.rs

```diff
@@ -0,0 +1,501 @@
+        for rack in self.racks.iter_mut() {
+            if racks.iter().any(|&new_rack| rack.name == new_rack.name()) {
+                rack.is_old = false;
+            }
+            rack.mark_new(disks);
+        }
+        self.racks.sort_unstable_by_key(|a| match a.is_old {
+            true => 0,
+            false => 1,
+        });
+    }
+
+    pub fn push(&mut self, item: Rack) {
+        self.racks.push(item)
+    }
+
+    pub fn disks_count(&self) -> usize {
+        self.racks.iter().fold(0, |acc, r| acc + r.disks_count())
+    }
+
+    fn next_disk(&self) -> Option<(&Node, &Disk)> {
+        let disks = self.racks.iter().flat_map(|r| {
+            r.nodes
+                .iter()
+                .flat_map(|n| n.disks.iter().map(move |d| (n, d)))
+        });
+        let (node, disk) = disks.min_by_key(|r| r.1.used_count.load(ORD))?;
+        node.inc();
+        disk.inc();
+        Some((node, disk))
+    }
+
+    fn next_disk_from_list(&self, list: &[Replica]) -> Option<(&Node, &Disk)> {
+        let mut disks = self.racks.iter().flat_map(|r| {
+            r.nodes
+                .iter()
+                .flat_map(|n| n.disks.iter().map(move |d| (n, d)))
+        });
+
+        let min_key = disks.clone().map(|x| x.1.used_count.load(ORD)).min()?;
+
+        let (node, disk) = disks.find(|(node, disk)| {
+            disk.used_count.load(ORD) == min_key
+                && list.contains(&Replica::new(node.name.clone(), disk.name.clone()))
+        })?;
+
+        node.inc();
+        disk.inc();
+        Some((node, disk))
+    }
+
+    fn next_rack(&self) -> Option<&Rack> {
+        let rack = self
+            .racks
+            .iter()
+            .min_by_key(|r: &&Rack| r.used_count.load(ORD))?;
+        rack.inc();
+        Some(rack)
+    }
+
+    fn next_rack_from_list(&self, list: &[&str]) -> Option<&Rack> {
+        if list.is_empty() {
+            return None;
+        }
+        let min_key = self.racks.iter().map(|x| x.used_count.load(ORD)).min()?;
+
+        let rack = self.racks.iter().find(|&rack| {
+            rack.used_count.load(ORD) == min_key && list.contains(&rack.name.as_str())
+        })?;
+
+        rack.inc();
+        Some(rack)
+    }
+
+    pub fn create_vdisk(&self, id: u32, replicas_count: usize) -> VDisk {
+        let mut vdisk = VDisk::new(id);
+        let first_rack = self.next_rack().expect("no racks in setup");
+        let (node, disk) = first_rack.next_disk(&[]).expect("no disks in setup");
+        vdisk.push_replica(Replica::new(node.name.clone(), disk.name.clone()));
+
+        while vdisk.replicas().len() < replicas_count {
+            let banned_nodes = get_used_nodes_names(vdisk.replicas());
+            let (node, disk) = if vdisk.replicas().len() == 2 {
+                first_rack.next_disk(&banned_nodes).map_or_else(
+                    || {
+                        self.next_rack()
+                            .expect("no racks in setup")
```

Comment:
You should make "new" method accept racks, and check there it's emptyness. This way you can omit "Option" in "next_" methods, as you'll be sure it's not empty

Reply:
And "new" should  return Result<..Center>, so you can log error